### PR TITLE
A clone of a Func should share cache entries with the original

### PR DESCRIPTION
**Pull request by abadams**
This permits some interesting dynamic reuse possibilities within a single pipeline.

This required reworking the memoize instance counter to be per-lowering instead of per-func, and tracking the name of the original Func through clones to use in the cache key.

@zvookin @jrk

### CODE REVIEW

**Comment by abadams** [comment]
One downside of this is that it requires the multiple uses to be exactly the same size. We may want to consider relaxing the cache to permit hits on subsets of buffers. May require device_crop?

File: src/Function.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ class WeakenFunctionPtrs : public IRMutator2 {
 }
 
 struct FunctionContents {
-    std::string name;
+    std::string name, parent_name;
```
Comment (zvookin) [inline src/Function.cpp]: Is there a better name for this? "origin_name" perhaps? Also, separate lines seems better.

File: src/Function.h [hunk 2]
```diff
@@ -137,6 +137,11 @@ class Function {
     /** Get the name of the function. */
     EXPORT const std::string &name() const;
 
+    /** If this is a wrapper of another func, created by a chain of in
+     * or clone_in calls, returns the name of the original
+     * Func. Otherwise returns the name. */
+    EXPORT const std::string &parent_name() const;
```
Comment (zvookin) [inline src/Function.h]: So here's a question I don't know the answer to: Do we want to just track the name, or do we want to return the Func?

File: test/correctness/clone_cache_key.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#include "Halide.h"
```
Comment (zvookin) [inline test/correctness/clone_cache_key.cpp]: It would be more descriptive if this filename contained "memoize" e.g. "memoize_cloned.cpp".

**Comment by zvookin** [comment]
Per the comment on the uses needing to be the same size, this is a general issue in memoize right? I.e. calling a filter in two ways where one computes a subset of the other will result in two memoize entries. It would be great to address this generally.

Re: device_crop, did we ever make the memoize cache handle device allocations at all? This would be another thing to improve. Though doing so definitely requires increasing client control of cache size, e.g. per device type, and adding richer ways to clear specific entries.

**Pull request merged**